components/script/dom/vr.rs

```diff
@@ -128,34 +131,112 @@ impl VR {
+                        if let Some(display) = self.find_display(id) {
+                            display.handle_webvr_event(&event);
+                        self.notify_event(&display, &event);
+                        }
+                    },
+                    &WebVRDisplayEvent::Activate(ref display, _) |
+                    &WebVRDisplayEvent::Deactivate(ref display, _) |
+                    &WebVRDisplayEvent::Blur(ref display) |
+                    &WebVRDisplayEvent::Focus(ref display) |
+                    &WebVRDisplayEvent::PresentChange(ref display, _) |
+                    &WebVRDisplayEvent::Change(ref display) => {
+                        let display = self.sync_display(&display);
+                        display.handle_webvr_event(&event);
+                    }
+                };
             },
-            &webvr::VRDisplayEvent::Activate(ref display, _) |
-            &webvr::VRDisplayEvent::Deactivate(ref display, _) |
-            &webvr::VRDisplayEvent::Blur(ref display) |
-            &webvr::VRDisplayEvent::Focus(ref display) |
-            &webvr::VRDisplayEvent::PresentChange(ref display, _) |
-            &webvr::VRDisplayEvent::Change(ref display) => {
-                let display = self.sync_display(&display);
-                display.handle_webvr_event(&event);
+            WebVREventMsg::GamepadEvent(event) => {
+                match &event {
+                    &WebVRGamepadEvent::Connect(ref data, ref state) => {
+                        if let Some(gamepad) = self.find_gamepad(state.gamepad_id) {
+                            gamepad.update_from_vr(&state);
+                        } else {
+                            // new gamepad
+                            self.sync_gamepad(Some(data.clone()), state);
+                        }
+                    },
+                    &WebVRGamepadEvent::Disconnect(id) => {
+                        if let Some(gamepad) = self.find_gamepad(id) {
+                            gamepad.update_connected(false);
+                        }
+                    }
+                };
             }
         };
     }
 
-    fn notify_event(&self, display: &VRDisplay, event: &webvr::VRDisplayEvent) {
+    pub fn handle_webvr_events(&self, events: Vec<WebVREventMsg>) {
+        for event in events {
+            self.handle_webvr_event(event);
+        }
+    }
+
+    fn notify_event(&self, display: &VRDisplay, event: &WebVRDisplayEvent) {
         let event = VRDisplayEvent::new_from_webvr(&self.global(), &display, &event);
         event.upcast::<Event>().fire(self.upcast());
     }
 }
 
+// Gamepad
+impl VR {
+    fn find_gamepad(&self, gamepad_id: u64) -> Option<Root<Gamepad>> {
+        self.gamepads.borrow()
+                     .iter()
+                     .find(|g| g.gamepad_id() == gamepad_id)
+                     .map(|g| Root::from_ref(&**g))
+    }
+
+    fn sync_gamepad(&self, data: Option<WebVRGamepadData>, state: &WebVRGamepadState) {
```

Comment:
Take a `Option<&WebVRGamepadData>` here, this avoids a cloning operation when handling `WebVRGamepadEvent::Connect` above.